packages/roosterjs-editor-core/lib/undo/snapshotUtils.ts

```diff
@@ -0,0 +1,139 @@
+import Editor from '../editor/Editor';
+import browserData from '../utils/BrowserData';
+import { EditorPoint, NodeBoundary, NodeType } from 'roosterjs-editor-types';
+import { normalizeEditorPoint } from 'roosterjs-editor-dom';
+
+// Undo cursor marker
+const CURSOR_START = 'cursor-start';
+const CURSOR_END = 'cursor-end';
+
+// Build undo snapshot
+export function buildSnapshot(editor: Editor): string {
+    // Build the snapshot in-between adding and removing cursor marker
+    let selectionRange = editor.getSelectionRange();
+    if (selectionRange) {
+        addCursorMarkersToSelection(editor, selectionRange);
+    }
+
+    let htmlContent = editor.getContent(false /*triggerExtractContentEvent*/) || '';
+
+    // This extra update selection to cursor marker post building snapshot is added for Mac safari
+    // We temporarily inject a cursor marker to current selection prior to build snapshot and remove it afterwards
+    // The insertion of cursor marker for some reasons has caused the selection maintained in browser to be lost.
+    // This restores the selection prior to removing the cursor marker.
+    // The code may throw error for Firefox and IE, hence keep it only for Mac Safari
+    if (browserData.isSafari) {
+        updateSelectionToCursorMarkers(editor);
+    }
+
+    removeCursorMarkers(editor);
+    return htmlContent;
+}
+
+// Restore a snapshot
+export function restoreSnapshot(editor: Editor, snapshot: string): void {
+    editor.setContent(snapshot);
+
+    // Restore the selection and delete the cursor marker afterwards
+    updateSelectionToCursorMarkers(editor);
+    removeCursorMarkers(editor);
```

Comment:
I've seen cases where cursor marker still exists after some actions. Maybe errors are thrown in 'updateSelectionToCursorMarkers', should we make it like 'try{} finally{}'?